### PR TITLE
floats: add ParseWithNA

### LINKED ISSUE #44: proposal: add ParseFloat64(s string) (v, w float64, err error)
We provide weighted value support for tests etc in stat, but there is no handling for textual representation of "NA" provided by stdlib. So I propose we add in either stat or floats a `ParseFloat` or `Parse` respectively that handles this.

Edited to incorporate suggestion below to allow a user-provided na string.

Assume in floats.
```
// Parse converts the string s to a float64 in v. If s equals na, w is returned as 0, otherwise 1.
func Parse(s, na string) (v, w float64, err error) {
    if s == na {
        return 0, 0, nil
    }
    v, err = strconv.ParseFloat(s, 64)
    if err == nil {
        w = 1
    }
    return v, w, err
}
```

This would then be used like so, e.g. single column data:
```
var vals, weights []float64
sc := bufio.NewScanner(r)
for sc.Scan() {
    v, w, err := floats.Parse(sc.Text(), "NA")
    if err != nil { ... }
    vals = append(vals, v)
    weights = append(weights, w)
}
err := sc.Err() ...
mean := stat.Mean(vals, weights)
```

### PR DESCRIPTION
**Pull request by kortschak**
@btracey Please take a look.

Closes #44.

### CODE REVIEW

File: floats/floats.go [hunk 1]
```diff
@@ -567,6 +568,19 @@ func Norm(s []float64, L float64) float64 {
 	return math.Pow(norm, 1/L)
 }
 
+// ParseWithNA converts the string s to a float64 in v.
+// If s equals na, w is returned as 0, otherwise 1.
```
Comment (btracey) [inline floats/floats.go]: Can we somehow use the phrase "not available" into the documentation? I feel the purpose isn't very clear, especially for non-native speakers.

File: floats/parse_example_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/stat"
+)
+
+func ExampleParseWithNA() {
+	const data = `6
+NA
+4
+`
+
```
Comment (btracey) [inline floats/parse_example_test.go]: It would be nice if this had some comment like 
"Compute the mean of a list of numbers, ignoring missing entries"

**Comment by kortschak** [comment]
PTAL

File: floats/floats.go [hunk 3]
```diff
@@ -567,6 +568,19 @@ func Norm(s []float64, L float64) float64 {
 	return math.Pow(norm, 1/L)
 }
 
+// ParseWithNA converts the string s to a float64 in v.
```
Comment (btracey) [inline floats/floats.go]: Should this be ParseWithMissing ? NA is shorter, but missing reads much better.

Comment (kortschak) [inline floats/floats.go]: I was thinking about that and I think the shortness should win here. It's almost never the case that the word "missing" is used, but "NA" is very common. I'll wait for other input.

**Comment by btracey** [comment]
@dwhitena

File: floats/floats.go (hunk 3 above)
Comment (vladimir-ch) [inline floats/floats.go]: Recently I read an article about NaN and NA. I had to look up what NA means (but on the other hand I never processed any external data). Now I would side with ParseWithNA.

**Comment by kortschak** [comment]
Was just checking out how R handles NA and I find this:

```
typedef union
{
    double value;
    unsigned int word[2];
} ieee_double;
```

```
static double R_ValueOfNA(void)
{
    volatile ieee_double x;
    x.word[hw] = 0x7ff00000;
    x.word[lw] = 1954;
    return x.value;
}
```

So they are using a specific `NaN` ([`0x7ff80000000007a2`](https://play.golang.org/p/7ntbMNCD5v)) as NA. According to [this](http://adv-r.had.co.nz/C-interface.html), the 1954 comes from one of the original R authors' birthdays. This approach is more space efficient than what we have but would involve significant rewriting of routines for them to be NA-aware.

**Comment by btracey** [comment]
It's a clever idea, but I think it'll be hard to operationalize throughout the suite. Ignoring NA values (by setting 0 weight) is one reasonable thing to do, but for many problems that's not the correct thing to do. Maybe NA values arise because the monitoring period was only so long, and so NA really means `> 5 weeks` .  I see handling NA values as part of the steps between loading values and performing computation. It seems to me to be better to force that handling step early, rather than having every `gonum` routine have a set of complicated actions and options around NA .

**Pull request merged**